### PR TITLE
[WIP] Towards Issue #875, avoid duplicating files added to ipfs

### PR DESCRIPTION
**Pull request by kevina**
**NOT READY FOR MERGE**

This is a work in progress, but I wanted to get some early feedback on my work towards #875.  Implementing this feature touched a lot of code and requires some API changes.

Basic adding of files without storing the data now works.  This currently needs to be done when the node is offline.  To use:
```
add --no-copy <some local file>
```
will add the file without copying the blocks.  If the file is moved or changed than any blocks created from that file will become invalid.

Notes for review:
- The first two commits will not be in the final full request.  The commit "Refactor Makefile" has been submitted separately and the commit "Temporary disable a failing test" should be dealt with separately.
- All of the commits with "Required for #875" mean what the commit says and **are the ones that I want early feedback on right now**.
- Some of the commits directly modify go-datastore in the Godeps/ directory.  I did this for simplicity and also taked to @whyrusleeping on irc about this.  This will be cleaned up before the final pull request.

### CODE REVIEW

**Comment by jefft0** [comment]
Hi Kevin. Has someone looked at your pull request? Do you need someone to try it? (Or wait for it to be more baked?)

**Comment by whyrusleeping** [comment]
I've looked over a bit of it, holding off on leaving any comments or doing a serious review until its asked for

**Comment by kevina** [comment]
@whyrusleeping Please do a serious review on what functions keeping in mind this is a work in progress.  Here is what works as of now:

* `ipfs add --nocopy FILE` to add a file without copying in offline mode.  If the server is not offline you will get a error about a non-absolute pathname.
* `ipfs filestore ls` list files in the filestore
* `ipfs filestore verify` verify the contents in the filestore.

There is still lots to do, but I am hesitant on putting to much work until I get some feedback that I am heading down the right path.

@jefft0 yes please try it out

Both:  Commit up to (and including) 01081e7 is stable and I won't do any forced updates until this is closer to getting into master.  Commits after that are newer and slightly less stable and I might do a forced update if I discover a problem (for example tests failing).  If this is a problem for either of you let me know.

**Comment by whyrusleeping** [comment]
This is looking promising, there are definitely a few problems we're going to have to think through, but this is progress.

The first thing I suggest is to not add an extra parameters to the blockstore methods, instead we can look at making `Block` an interface, and having the current implementation be `rawBlock` or something, and then your `filedataBlock` can also implement it. We can do type switching internally to handle things differently.

**Comment by kevina** [comment]
@whyrusleeping I am not sure I am completely following.  Are you saying you would like to see something like the `filestore.DataWOpts` for blocks also and have `filedataBlock` also contain the add options to avoid adding the `addOpts` paramater to the blockstore methods?

**Comment by whyrusleeping** [comment]
@kevina yeah, basically. The idea is also to not have to do weird augments to the types like youre doing to `Block`. Instead, make `Block` an interface, and create a completely separate type for what youre doing.

**Comment by kevina** [comment]
@whyrusleeping I see what you are saying about not adding the `DataPtr` field to `Block`.

Are you also against adding a generic `addOpts` parameter to the blockstore methods as a way of passing additional parameters when adding blocks?  The `addOpts` parameter is a generic `interface{}` and can simply be set to `nil` when not used.  Having a way to pass generic parameters to the `Add()` and `Put()` methods seams like a good thing to have going forward.    In fact, I was going to propose that the datastore methods also accept a generic `opts` paramater, but it would of touched too much code and would of changed the `Datastore` interface.

In the future I might want a way to pass additional parameters to the `Get()` methods.  With the `Add()` and `Put()` methods I can hack option passing by including the parameters with the data object to be added (like I did with `filestore.DataWOpts`). With the Get() methods such as hack is not possible unless you also make `Key` an interface.

**Comment by whyrusleeping** [comment]
I am also against adding extra parameters to the existing methods. I think keeping the methods simple and using the type system for conveying extra information is a better option.

**Comment by kevina** [comment]
@whyrusleeping okay.  I just pushed two commits to make blocks.block an interface and remote the addOpts parameter from blockstore methods.

File: repo/repo.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ type Repo interface {
 	// SetAPIAddr sets the API address in the repo.
 	SetAPIAddr(addr string) error
 
+	Self() Repo
```
Comment (whyrusleeping) [inline repo/repo.go]: why is this needed?

File: importer/chunk/splitting.go [hunk 2]
```diff
@@ -12,7 +13,10 @@ var log = logging.Logger("chunk")
 var DefaultBlockSize int64 = 1024 * 256
 
 type Splitter interface {
-	NextBytes() ([]byte, error)
+	// returns the data, an offset if applicable and, an error condition
+	NextBytes() ([]byte, int64, error)
```
Comment (whyrusleeping) [inline importer/chunk/splitting.go]: I think instead of adding an offset here to the return (which will pretty much only be used by this codepath). We should change the `[]byte` return to be some type that allows us to either return data, or to return a file piece, similar to the change made to the block object. This new type (whatever we end up making) can also replace the data field in the merkledag.Node object to simplify the `DataPtr` stuff you've got going.

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -50,7 +50,8 @@ func (i *gatewayHandler) newDagFromReader(r io.Reader) (*dag.Node, error) {
 	// return ufs.AddFromReader(i.node, r.Body)
 	return importer.BuildDagFromReader(
 		i.node.DAG,
-		chunk.DefaultSplitter(r))
+		chunk.DefaultSplitter(r),
+		nil)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: if we make the splitter change i mentioned earlier, we can move the 'add opts' into the constructor for the splitter, instead of this call.

File: Godeps/_workspace/src/github.com/ipfs/go-datastore/measure/measure.go [hunk 4]
```diff
@@ -179,10 +179,9 @@ func (m *measure) Batch() (datastore.Batch, error) {
 func (mt *measuredBatch) Put(key datastore.Key, val interface{}) error {
 	mt.puts++
 	valb, ok := val.([]byte)
-	if !ok {
-		return datastore.ErrInvalidType
+	if ok {
+		_ = mt.m.putSize.RecordValue(int64(len(valb)))
```
Comment (whyrusleeping) [inline Godeps/_workspace/src/github.com/ipfs/go-datastore/measure/measure.go]: Here youre starting to touch on a change i've wanted to make to the datastore for a while now.

Instead of just throwing the length away if the value isnt bytes, how about we check if the value has a `Len()` method?

We could define an interface (hrm, this sounds familiar) that has something like:
```
type Value interface {
  Data() []byte
  Len() int
}
```
and then just type check for that object, and use its `Len()` method here if available.

The awesome thing this would let us do is pass the `Block` object from the blockstore *directly* into the datastore.Put. Once we get that working, we can extend the interface differently down the road (example: maybe something like a `WriteTo` method that would help us avoid keeping the blocks in memory)

File: repo/repo.go (hunk 1 above)
Comment (kevina) [inline repo/repo.go]: I needed a way to be able to get to the `FSRepo` and hence get to the `filestore.Datastore` in the next commit (4ef5531).  In this commit `node.Repo.(*fsrepo.FSRepo)` does not work as node.Repo is `repo.ref` so I needed to use add the Self() method and use `node.Repo.Self().(*fsrepo.FSRepo)`

File: Godeps/_workspace/src/github.com/ipfs/go-datastore/measure/measure.go (hunk 4 above)
Comment (kevina) [inline Godeps/_workspace/src/github.com/ipfs/go-datastore/measure/measure.go]: @whyrusleeping I somewhat agree, but this change will touch a lot of code and change the datastore interface.  Since it doesn't seam necessary for getting my changes in can we agree to hold of on this until this pull request is ready to merge?

File: repo/repo.go (hunk 1 above)
Comment (whyrusleeping) [inline repo/repo.go]: the `node.Repo` should either be an fsrepo or a mock repo. It should work with a simple type assertion

> In this commit `node.Repo.(*fsrepo.FSRepo)` does not work as node.Repo is `repo.ref`

I'm not sure what you mean by this. what is repo.ref?

Comment (kevina) [inline repo/repo.go]: `node.Repo` is _not_ a `FSRepo` it is a `repo.ref`.  `repo.ref` is defined here: https://github.com/ipfs/go-ipfs/blob/c067fb9e83e89cf04226d2c43de7c6fd5ebbccd2/repo/onlyone.go#L50.  It may be easier for you to just try it without the Self() and see for yourself.

Comment (whyrusleeping) [inline repo/repo.go]: Ah, I see now. I incorrectly assumed the fsrepo.Open constructor returned an fsrepo

**Comment by kevina** [comment]
@jefft0 if you want to do some preliminary testing now would be a good time

Here is what you can try:

Add a file with `ipfs add --no-copy`.  Try to download the hash from a different ipfs node (such as the gateway).  Look at the contents of the filestore with `ipfs filestore ls`.  Change the contents of a file or move a file and run `ipfs filestore verify`.  Use `ipfs filestore rm-invalid` to clean up any invalid objects due to the file changing.

There is basic help now available for the `ipfs filestore` commands, please try it out and let me know if my text makes sense.

Objects are pinned as usual when adding, but removing the pin and running `ipfs repo gc` will not remove the object from the filestore.  That is because I am unsure if this is a desirable thing to do as the block stubs take very little space.  (It is also unclear if filestore objects should be pinned at all.)  So for now objects need to be removed manually with `ipfs filestore rm <hash>`.  This command will also unpin the file if necessary.

**Comment by jefft0** [comment]
Boy, that's a lot of work in a short amount of time! `add --no-copy` and the `filestore` comands work for me. The `--help` is useful and makes sense to me. 

I found one counter-intuitive thing. I made a short hello file and did `ipfs add --no-copy hello.txt`. After adding more text to the end of the file, `ipfs filestore verify` still reports OK. I understand that filestore saves a slice and the slice is still good. But I want to get the hashes of the files I have on disk and ask IPFS if it has them. Now the hashes are different. Do you see the issue? What do you think? Is there a middle-ground solution?

**Comment by kevina** [comment]
@jefft0 Thanks for testing it out!

> Do you see the issue? What do you think? Is there a middle-ground solution?

I don't see it as a major issue. The objects are actually still valid, if you try to get the old object the contents would not have changed.  "ipfs filestore verify" is meant to verify individual blocks not complete files.  It should be fairly easy to add a command that will verify files and check that the "WholeFile" flag is correct, but that is a low priority for me right now.

**Comment by jefft0** [comment]
OK. Is it ready for me to stress test with a bunch of 200 MB video files (my use case)? If you're still tweaking performance, I'll hold off.

**Comment by kevina** [comment]
> OK. Is it ready for me to stress test with a bunch of 200 MB video files (my use case)? If you're still tweaking performance, I'll hold off.

I have not tuned it for performance at all but I would be interested in how well it performs on a large file.

**Comment by kevina** [comment]
@whyrusleeping 

> I think instead of adding an offset here to the return (which will pretty much only be used by this codepath). We should change the []byte return to be some type that allows us to either return data, or to return a file piece, similar to the change made to the block object. This new type (whatever we end up making) can also replace the data field in the merkledag.Node object to simplify the DataPtr stuff you've got going.

I just pushed some commits that basically implemented this and simplified a lot of code.  The DataPtr stuff in merkledag.Node contains different information so I can not use this new type for that.

**Comment by jefft0** [comment]
> I would be interested in how well it performs on a large file.

On my MacBook Pro, I added a folder with 24 video files totaling 3GB. `ipfs add -r` takes 33 seconds and `ipfs add -r --no-copy` only takes 18 seconds.

**Comment by vitzli** [comment]
I found weird bug/glitch in `filestore verify/ls`: it outputs the hash that does not start with `Qm…` (current version), could somebody verify it?
Environment: Debian Stretch, current version, amd64, recent version from kevina/go-ipfs, branch issue-875-wip (built couple hours ago).

It happened when I tried to add/verify [ceph-base_10.2.0-1~bpo80%2b1_amd64.deb](http://download.ceph.com/debian-jewel/pool/main/c/ceph/ceph-base_10.2.0-1~bpo80%2b1_amd64.deb), ([copy](https://www.dropbox.com/s/ujbunl9t012ompz/ceph-base_10.2.0-1~bpo80%2B1_amd64.deb?dl=0) for archival reasons):

```
user@deb-stretch:~$ ipfs add --no-copy ceph-base_10.2.0-1~bpo80+1_amd64.deb 
added QmU1bhY1HN4QScSnJNQRQ5qmnpwfZMA3bDs2fhdCkz2rkr ceph-base_10.2.0-1~bpo80+1_amd64.deb

same with 

user@deb-stretch:~$ ipfs add -n ceph-base_10.2.0-1~bpo80+1_amd64.deb 
added QmU1bhY1HN4QScSnJNQRQ5qmnpwfZMA3bDs2fhdCkz2rkr ceph-base_10.2.0-1~bpo80+1_amd64.deb
```
but `ipfs filestore ls` produces:
```
…
QmXwJZRKWpbEDCsxAq7N4jT5Wu7N78HgLnwFn9WhMN3GvM leaf  /home/user/ceph-base_10.2.0-1~bpo80+1_amd64.deb 38797312 262144
6PKoQ1TrGRFrN7Ch9t9fQUmFLEYHqa7r3YhtrPU9eYvKd leaf  /home/user/ceph-base_10.2.0-1~bpo80+1_amd64.deb 34865152 262144
QmYCKEbjsC6VEWmZp22dVxUJnNYhWv2R2xDU9CeZ9zmrQ8 leaf  /home/user/ceph-base_10.2.0-1~bpo80+1_amd64.deb 45350912 262144
…
```
`verify` command return "ok" for this block.

The block could be extracted with `dd` (offset 34865152 is skip=133 with bs=262144):
```
dd if=ceph-base_10.2.0-1~bpo80+1_amd64.deb of=badblock-6PKo.bin skip=133 bs=262144 count=1
```
[badblock-6PKo.bin](https://www.dropbox.com/s/s47iu1cnctwfu4k/badblock-6PKo.bin?dl=0) - extracted block

Doing `add -n badblock-6PKo.bin` gives:
```
added QmXynZ3WT44YtANcECmEWdYJFtm7KPBDUEwVrzAqAjtMCi badblock-6PKo.bin
```
and `add --no-copy badblock-6PKo.bin` gives the same result.

**BUT**, doing `ipfs filestore ls` on it returns the same weird hash:
```
user@deb-stretch:~$ ipfs filestore ls
6PKoQ1TrGRFrN7Ch9t9fQUmFLEYHqa7r3YhtrPU9eYvKd leaf  /home/user/badblock-6PKo.bin - 262144
```
For `filestore rm`: doing `rm 6PKo…` removes the block without "unpinnig" text, but block disappears from `ls` output, while `filestore rm QmXyn…` says "unpinng" and removes the block from `ls/verify`.

**Comment by kevina** [comment]
@vitzli this is due to issue #2601 and #911.  It is very annoying but nothing I can control directly.  Please direct discussion to one of those bug reports.  If this becomes a major problem I am thinking about hacking around the problem in my code, but am holding off for now.

**Comment by kevina** [comment]
@jefft0.  I am glad to here that adding it is faster.  Retrieving locally is a bit slower, but by informal test have determined that is due to always verifying the hash.  A better solution might be use modification times and only verify when the file's modification timestamp has changed.

**Comment by jefft0** [comment]
@kevina, If I already did `ipfs add cats.mp4` and then do `ipfs add --no-copy cats.mp4` should it replace the copied values in the block store with links to `cats.mp4` (maybe to recover space in the block store)? Conversely, if I already did `ipfs add --no-copy cats.mp4` and then do `ipfs add cats.mp4` it will add the blocks to the block store. In this case, does `ipfs get` read blocks from the block store (ignoring the links to `cats.mp4`)?

File: blocks/blockstore/write_cache.go [hunk 5]
```diff
@@ -34,28 +34,36 @@ func (w *writecache) Has(k key.Key) (bool, error) {
...
 }
 
-func (w *writecache) PutMany(bs []*blocks.Block) error {
-	var good []*blocks.Block
+func (w *writecache) PutMany(bs []blocks.Block) error {
+	var good []blocks.Block
 	for _, b := range bs {
-		if _, ok := w.cache.Get(b.Key()); !ok {
+		// Don't cache "advance" blocks
+		if _, ok := b.(*blocks.BasicBlock); ok {
```
Comment (whyrusleeping) [inline blocks/blockstore/write_cache.go]: This check doesnt need to be just for a single block type. If we have a given block in the flatfs store, theres no need to add a filestore reference to it as well.

File: blocks/blockstore/blockstore.go [hunk 6]
```diff
@@ -89,37 +90,62 @@ func (bs *blockstore) Get(k key.Key) (*blocks.Block, error) {
...
 			return err
 		}
 	}
 	return t.Commit()
 }
 
+func (bs *blockstore) prepareBlock(k ds.Key, block blocks.Block) interface{} {
+	if fsBlock, ok := block.(*blocks.FilestoreBlock); !ok {
+		// Has is cheaper than Put, so see if we already have it
+		exists, err := bs.datastore.Has(k)
```
Comment (whyrusleeping) [inline blocks/blockstore/blockstore.go]: this check should be moved out of the type switch

File: commands/files/file.go [hunk 7]
```diff
@@ -11,6 +11,57 @@ var (
 	ErrNotReader    = errors.New("This file is a directory, can't use Reader functions")
 )
 
+// An AdvReader is like a Reader but supports getting the current file
+// path and offset into the file when applicable.
+type AdvReader interface {
+	io.Reader
+	ExtraInfo() ExtraInfo
```
Comment (whyrusleeping) [inline commands/files/file.go]: I don't see where all this `ExtraInfo` stuff is getting used. It seems like we're unnecesarily bloating the interfaces

**Comment by whyrusleeping** [comment]
I think the next step here is going to be to break up the changes in this PR and start working some of the 'framework' stuff into the codebase. For example: the change to the blocks.Block type could be extracted and merged in separately

File: blocks/blockstore/blockstore.go (hunk 6 above)
Comment (kevina) [inline blocks/blockstore/blockstore.go]: @whyrusleeping can you be more specific on what you want too see?

Comment (whyrusleeping) [inline blocks/blockstore/blockstore.go]:
```go
exists, err := bs.datastore.Has(k)
if err == nil && exists {
  return nil //already stored
}
switch block.(type) {
default:
  return block.Data()
case *blocks.FilestoreBlock:
  // Logic thats currently in your 'else' block
}
```

**Comment by kevina** [comment]
> @kevina, If I already did ipfs add cats.mp4 and then do ipfs add --no-copy cats.mp4 should it replace the copied values in the block store with links to cats.mp4 (maybe to recover space in the block store)?

At the moment the blocks will be in both datastores.  I am working on doing something about this.

**Comment by kevina** [comment]
> I think the next step here is going to be to break up the changes in this PR and start working some of the 'framework' stuff into the codebase. For example: the change to the blocks.Block type could be extracted and merged in separately

Okay, I will start by separating out the block.Block change to an interface type as I would like to see that go in to prevent bitrot.

File: blocks/blockstore/blockstore.go (hunk 6 above)
Comment (kevina) [inline blocks/blockstore/blockstore.go]: @whyrusleeping if a block is already in the normal datastore and a user adds the file with "--nocopy" I want the file to be added to the filestore so that the block in the normal datastore can eventually be removed to save space.  That is why when we have a FilestoreBlock I skip the check Has() check.  I hope this makes sense.

**Comment by whyrusleeping** [comment]
> as I would like to see that go in to prevent bitrot.
100% agreed, avoiding stagnation is good

File: blocks/blockstore/blockstore.go (hunk 6 above)
Comment (whyrusleeping) [inline blocks/blockstore/blockstore.go]: Hrm... I disagree with that. It breaks abstractions. If users already have a block (regardless of how its stored) we don't want to store it again.

File: commands/files/file.go (hunk 7 above)
Comment (kevina) [inline commands/files/file.go]: @whyrusleeping , It is used to pass needed information through the layers into the filestore such as the absolute pathname and the offset into the file.  It is also used to pass addOpts through the lawyers to avoid that extra parameter.  Are you looking for specific spots in the code where the information is used?

Comment (whyrusleeping) [inline commands/files/file.go]: hrm... I'll look at this in more detail once we get the blocks changes in and this PR rebased on top of that.

File: blocks/blockstore/blockstore.go (hunk 6 above)
Comment (kevina) [inline blocks/blockstore/blockstore.go]: This is going to take some time to work through.  Right now I require certain information to be stored in the filestore, regardless of if the blocks are stored elsewhere.  Let's come back to this.

Comment (whyrusleeping) [inline blocks/blockstore/blockstore.go]: agreed, can discuss later

File: blocks/blockstore/write_cache.go (hunk 5 above)
Comment (kevina) [inline blocks/blockstore/write_cache.go]: The reason for this change is related to https://github.com/ipfs/go-ipfs/pull/2600#discussion_r62229755

**Comment by kevina** [comment]
Closing and creating new request, See #2634.
